### PR TITLE
#3469: Add support for stack usage reporting through Watcher

### PR DESCRIPTION
**Pull request by tt-dma**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/3469

### What's changed
New watcher feature to track stack usage, updated after each kernel and read out from mailboxes by watcher server.

### Checklist
- [x] Post commit CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/10259739052
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable) - N/A
- [x] New/Existing tests provide coverage for changes - Updated run_tools_tests.sh

### CODE REVIEW

File: tt_metal/hw/inc/debug/stack_usage.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    return;
+}
+
+void record_stack_usage() {
+    // Write the pause flag for this core into the memory mailbox for host to read.
+    debug_stack_usage_t tt_l1_ptr *stack_usage_msg = GET_MAILBOX_ADDRESS_DEV(watcher.stack_usage);
+
+    uint32_t tt_l1_ptr *stack_ptr = (uint32_t tt_l1_ptr *) get_stack_base();
+    for (uint32_t stack_offset = 0; stack_offset < (get_stack_size() / sizeof(uint32_t)); stack_offset++) {
```
Comment (pgkeller) [inline tt_metal/hw/inc/debug/stack_usage.h:68]: confused, isn't this backwards?  stack grows high to low, you are starting low and going high, so we expect the first value to be the dirty pattern and we exit when we find not the dirty pattern ?

**Review by pgkeller — CHANGES_REQUESTED** [review]

File: tt_metal/hw/inc/debug/stack_usage.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#else
+    return 0;
+#endif
+}
+
+void dirty_stack_memory() {
+    uint32_t tt_l1_ptr *stack_ptr = (uint32_t tt_l1_ptr *) get_stack_base();
+    uint32_t stack_size = get_stack_size();
+    uint32_t stack_max_offset = stack_size / sizeof(uint32_t);
+    uint32_t stack_dirty_offset = stack_max_offset / 4;
```
[the remark this answers is not in the document]
Comment (tt-dma) [reply tt_metal/hw/inc/debug/stack_usage.h]: Yeah I had it backwards before 😬 , should be fixed now

File: tt_metal/hw/inc/debug/stack_usage.h (hunk 1 above)
Comment (tt-dma) [reply tt_metal/hw/inc/debug/stack_usage.h:72]: Should be fixed now

File: tt_metal/impl/debug/watcher_server.cpp [hunk 3]
```diff
@@ -555,6 +569,24 @@ static void dump_sync_regs(FILE *f, Device *device, CoreCoord core) {
...
 
+static void dump_stack_usage(FILE *f, Device *device, CoreCoord core, const mailboxes_t *mbox) {
+    for (int risc_id = 0; risc_id < DebugNumUniqueRiscs; risc_id++) {
+        uint16_t stack_usage = mbox->watcher.stack_usage.max_usage[risc_id];
+        uint16_t stack_size = get_riscv_stack_size(core, risc_id);
+        if (stack_usage != watcher::DEBUG_SANITIZE_NOC_SENTINEL_OK_16) {
+            fprintf(f, "\n\t%s stack usage: %d/%d", get_riscv_name(core, risc_id), stack_usage, stack_size);
+            if (stack_usage >= stack_size) {
+                fprintf(f, " (OVERFLOW)");
+                log_warning(
```
[the remark this answers is not in the document]
Comment (tt-dma) [reply tt_metal/impl/debug/watcher_server.cpp]: Added this in latest commit

File: tt_metal/impl/debug/watcher_server.cpp [hunk 4]
```diff
@@ -997,6 +1038,8 @@ void watcher_init(Device *device) {
                 device->id(), worker_core, debug_assert_init_val, GET_MAILBOX_ADDRESS_HOST(watcher.assert_status));
             tt::llrt::write_hex_vec_to_core(
                 device->id(), worker_core, debug_pause_init_val, GET_MAILBOX_ADDRESS_HOST(watcher.pause_status));
+            tt::llrt::write_hex_vec_to_core(
```
[the remark this answers is not in the document]
Comment (tt-dma) [reply tt_metal/impl/debug/watcher_server.cpp]: Figured best to add this now, it cleans things up quite a bit, added in it's own commit

**Comment by tt-dma** [comment]
CI looks good post-changes: https://github.com/tenstorrent/tt-metal/actions/runs/10272491524

**Review by pgkeller — APPROVED** [review]

**Review by aliuTT — APPROVED** [review]

**Pull request merged**